netmaster/objApi/apiController.go

```diff
@@ -0,0 +1,639 @@
+func (ac *APIController) ServiceInstanceDelete(serviceInstance *contivModel.ServiceInstance) error {
+	log.Infof("Received ServiceInstanceDelete: %+v", serviceInstance)
+	return nil
+}
+
+// TenantCreate creates a tenant
+func (ac *APIController) TenantCreate(tenant *contivModel.Tenant) error {
+	log.Infof("Received TenantCreate: %+v", tenant)
+
+	if tenant.TenantName == "" {
+		return errors.New("Invalid tenant name")
+	}
+
+	// Get the state driver
+	stateDriver, err := utils.GetStateDriver()
+	if err != nil {
+		return err
+	}
+
+	// Build tenant config
+	tenantCfg := intent.ConfigTenant{
+		Name:           tenant.TenantName,
+		DefaultNetType: "vlan",
+		SubnetPool:     tenant.SubnetPool,
+		AllocSubnetLen: uint(tenant.SubnetLen),
+		VLANs:          tenant.Vlans,
+		VXLANs:         tenant.Vxlans,
+	}
+
+	// Create the tenant
+	err = master.CreateTenant(stateDriver, &tenantCfg)
+	if err != nil {
+		log.Errorf("Error creating tenant {%+v}. Err: %v", tenant, err)
+		return err
+	}
+
+	// Create private network for the tenant
+	err = contivModel.CreateNetwork(&contivModel.Network{
+		Key:         tenant.TenantName + ":" + "privateNet",
+		IsPublic:    false,
+		IsPrivate:   true,
+		Encap:       "vxlan",
+		Subnet:      "10.1.0.0/16",
+		NetworkName: "privateNet",
+		TenantName:  tenant.TenantName,
+	})
+	if err != nil {
+		log.Errorf("Error creating privateNet for tenant: %+v. Err: %v", tenant, err)
+		return err
+	}
+
+	// Create public network for the tenant
+	err = contivModel.CreateNetwork(&contivModel.Network{
+		Key:         tenant.TenantName + ":" + "publicNet",
+		IsPublic:    true,
+		IsPrivate:   false,
+		Encap:       "vlan",
+		Subnet:      "192.168.1.0/24",
+		NetworkName: "publicNet",
+		TenantName:  tenant.TenantName,
+	})
+	if err != nil {
+		log.Errorf("Error creating publicNet for tenant: %+v. Err: %v", tenant, err)
+		return err
+	}
+
+	// Create a default volume profile for the tenant
+	err = contivModel.CreateVolumeProfile(&contivModel.VolumeProfile{
+		Key:               tenant.TenantName + ":" + "default",
+		VolumeProfileName: "default",
+		TenantName:        tenant.TenantName,
+		DatastoreType:     "none",
+		PoolName:          "",
+		Size:              "",
+		MountPoint:        "",
+	})
+	if err != nil {
+		log.Errorf("Error creating default volume profile. Err: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// TenantUpdate updates a tenant
+func (ac *APIController) TenantUpdate(tenant, params *contivModel.Tenant) error {
+	log.Infof("Received TenantUpdate: %+v, params: %+v", tenant, params)
+
+	return errors.New("Cant change tenant parameters after its created")
+}
+
+// TenantDelete deletes a tenant
+func (ac *APIController) TenantDelete(tenant *contivModel.Tenant) error {
+	log.Infof("Received TenantDelete: %+v", tenant)
+
+	// Get the state driver
+	stateDriver, err := utils.GetStateDriver()
+	if err != nil {
+		return err
+	}
+
+	// FIXME: Should we walk all objects under the tenant and delete it?
```

Comment:
I think we need to unlink the (network, service, app etc) objects here, so that they can go to unresolved state and trigger appropriate cleanups.

`DeleteTenantId()` used to delete the networks and endpoints under the tenant so I think in it's current form it will take care of cleaning up state from netmaster and netplugins. Volumes, epgs etc might need to be handled similar way.